scripts/core_uploader.sh

```diff
@@ -0,0 +1,97 @@
+
+
+# This script uploads the following to the provided S3 bucket under S3_KEY_PREFIX
+#   <s3_key>_<date>_HOSTNAME_RANDOM.core.zip
+#   <s3_key>_<date>_HOSTNAME_RANDOM.executable
+#   <s3_key>_<date>_HOSTNAME_RANDOM.stacktrace
+#   <s3_key>_<date>_HOSTNAME_RANDOM.all.zip
+
+
+export S3_BUCKET=$1
+export S3_KEY_PREFIX=$2
+export CRASH_FOLDER="/cores-out"
+
+# Folder Structure
+# $CRASH_FOLDER: All the crash symbol information is combined and added to this folder.
+#    It is uploaded to s3 and added to the /cores folder
+# /cores: Files are sent to the cores folder
+
+ls $CRASH_FOLDER/*
+
+# Process crash symbols
+export CORE_FILENAME=`basename $S3_KEY_PREFIX`_`date +"%FT%H%M%S"`${HOSTNAME+_host-$HOSTNAME}_${RANDOM_ID_VALUE}
+
+# Check if there is a corefile
+export NO_CORE_STR="ls: cannot access $CRASH_FOLDER/core*: No such file or directory"
+export LS_CORE_STR="$(ls $CRASH_FOLDER/core* 2>&1 >/dev/null)"
+
+if [ "$NO_CORE_STR" == "$LS_CORE_STR" ]; then
+
+    # No corefile
+    echo "No core file to upload"
+else
+    # Zip and upload the core file first incase the other processes fail
+    mv `ls $CRASH_FOLDER/core*` ${CORE_FILENAME}.core
+    zip $CRASH_FOLDER/${CORE_FILENAME}.core.zip $CRASH_FOLDER/${CORE_FILENAME}.core
+
+    # $CRASH_FOLDER
+    #   - CORE_FILENAME.core.zip
+    #   - CORE_FILENAME.core
+
+    # Upload core zip
+    aws s3 cp $CRASH_FOLDER/${CORE_FILENAME}.core.zip s3://${S3_BUCKET}/${S3_KEY_PREFIX}/
+
+    # Remove corefile zip to root folder
+    mv $CRASH_FOLDER/${CORE_FILENAME}.core.zip /tmp/${CORE_FILENAME}.core.zip
+
+    # $CRASH_FOLDER
+    #   - CORE_FILENAME.core
+
+    # Process corefile, get stack trace and fluent bit executable. Add to zip.
+    mv `ls $CRASH_FOLDER/core*` ${CORE_FILENAME}.core
+    cp /fluent-bit/bin/fluent-bit ./${CORE_FILENAME}.executable
+    gdb -batch -ex 'thread apply all bt full' -ex 'quit' '/fluent-bit/bin/fluent-bit' ${CORE_FILENAME}.core > $CRASH_FOLDER/${CORE_FILENAME}.stacktrace
+
+    # $CRASH_FOLDER
+    #   - CORE_FILENAME.core
+    #   - CORE_FILENAME.executable
+    #   - CORE_FILENAME.stacktrace
+
+    # Zip the crashfiles
+    zip -r $CRASH_FOLDER/${CORE_FILENAME}.all.zip $CRASH_FOLDER
+    mv /${CORE_FILENAME}.all.zip $CRASH_FOLDER/${CORE_FILENAME}.all.zip
+    rm $CRASH_FOLDER/${CORE_FILENAME}.core
+    mv /tmp/${CORE_FILENAME}.core.zip $CRASH_FOLDER/${CORE_FILENAME}.core.zip
+
+    # $CRASH_FOLDER
+    #   - CORE_FILENAME.core.zip
+    #   - CORE_FILENAME.executable
+    #   - CORE_FILENAME.stacktrace
+    #   - CORE_FILENAME.all.zip
+
+    # Send crash files to the /cores/ folder
+    cp $CRASH_FOLDER/* /cores/
+
+    # Send crash files to S3
+    aws s3 cp $CRASH_FOLDER s3://${S3_BUCKET}/${S3_KEY_PREFIX}/ --recursive
+
+    # Log stacktrace
+    echo "Stacktrace - ${CORE_FILENAME}.stacktrace:"
+    cat /cores/${CORE_FILENAME}.stacktrace
+
+    # Delay for 1 minute to allow for slow EFS uploads to complete if used
```

Comment:
explain EFS core file transfers?? once the AWS cp command is exited we've uploaded right?

Reply:
Right... So in my testing if the customer mounts an EFS volume to cores and you exit immediately, the core file doesn't actually get transfered to EFS... I needed to add a delay and then it seems like the volume gets synchronized

Reply:
interesintg. I guess I never actually tested that now that I think about... I only did the live debugging for running frozen task core file.

Reply:
Ah.... Yeah, a bit of an unexpected problem.